man/opendb.3
.TH OPENDB 3 "12 September 1995"
.FB
.SH NAME
opendb, closedb \- open/close a \fIFirstBase\fP database
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_opendb(db, mode, ixflag, ixoption)
.br
fb_database *db;
.br
int mode;
.br
int ixflag;
.br
int ixoption;
.sp 1
fb_closedb(db)
.br
fb_database *db;
.PP
.SH DESCRIPTION
As explained in the \fIdatabase\fP(4)
manual page, a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
\fIFb_opendb\fP and \fIfb_closedb\fP provide mechanisms to open and close an
entire
database \(em including simple and autoindex structures. In other words, the
entire database structure is initialized, with all needed file descriptors
and structure pointers.
\fIFb_opendb\fP will also provide file name extensions if needed.
.PP
The database structure needs to be allocated and the
file names need to be in the
database structure prior to the \fIfb_opendb\fP call. Typically, these
preliminary tasks are done using \fIdballoc\fP(3) and \fIdbargs\fP(3).
.PP
The \fImode\fP reflects desired permissions to the
database, with three possibilities \(em READ, WRITE, and READWRITE.
If a database does not exist, and the mode is WRITE or READWRITE, 
\fIfb_opendb\fP creates a database and database map, and initializes both.
.PP
The other two parameters are for database indexes. \fIIxflag\fP can take on
three values. FB_NOINDEX means that no index will be opened at all.
FB_WITHINDEX indicates that only the database index is to be opened.
FB_ALLINDEX means that all indexes, including the autoindexes, are to be
opened.
.PP
\fIIxoption\fP controls what \fIfb_opendb\fP will do if index errors occur
during the database initialization.
It also has three possible settings.
FB_OPTIONAL_INDEX means that no error messages are produced.
FB_MAYBE_OPTIONAL_INDEX will make error messages appear, but
control returns to the caller.
FB_MUST_INDEX will make \fIfb_opendb\fP die after displaying the
error message.
.PP
Fb_closedb closes all open file descriptors (databases, indexes, and the
database map), and frees up (most) all allocated space.
.PP
NOTE: There are allocated areas that fb_closedb does not free up
due to the underlying \fIFirstBase\fP structures for handling more than one
open database. To free and reset these areas, use \fIfree_globals\fP(3)
after all the databases have been closed using \fIfb_closedb\fP.
.SH EXAMPLES
Before opening a database from a C program, the database needs to be
defined using the \fIFirstBase\fP tool \fIdbdbas\fP(1).
Then, assuming the database is
named \fIPhone\fP, the following could be used to initialize and
open this database without any indexes:
.nf
.sp 1
.ft CW
   fb_database *db;
   ...
   db = fb_dballoc();
   fb_dbargs("Phone", NIL, db);
   fb_opendb(db, READWRITE, FB_NOINDEX, 0);
.ft
.sp 1
.fi
As another example, the following code fragment will set up the database
variables and open the database named \fIPhone\fP using an index
named \fIClients\fP:
.nf
.sp 1
.ft CW
   fb_database *db;
   ...
   db = fb_dballoc();
   fb_dbargs("Phone", "Clients", db);
   fb_opendb(db, READWRITE, FB_WITH_INDEX, FB_MUST_INDEX);
.ft
.sp 1
.fi
.sp 1
.fi
.SH FILES
.PD 0
.TP 10
dbase
default name of \fIFirstBase\fP database and screen.
.TP 10
index
default name of \fIFirstBase\fP index.
.TP 10
\\\\.cdb
extension of \fIFirstBase\fP database.
.TP 10
\\\\.map
extension of record map for \fIFirstBase\fP database.
.TP 10
\\\\.idx
extension of \fIFirstBase\fP database index.
.TP 10
\\\\.ddict
extension of \fIFirstBase\fP database dictionary.
.TP 10
\\\\.idict
extension of \fIFirstBase\fP database index dictionary.
.PD
.SH CLIENT/SERVER
The \fIfb_opendb\fP and \fIfb_closedb\fP routines will transparently
call the \fIfb_opendb_clnt\fP and \fIfb_closedb_clnt\fP
mechanisms when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP
is set to one. See \fIfb_clnt_create\fP(3) for more details.
.SH SEE ALSO
database(4), index(4), dballoc(3), dbargs(3), openidx(3), fb_free_globals(3).
.SH DIAGNOSTICS
If \fIfb_opendb\fP
cannot open the database file, it will die with an error message.
However, if a program sets the global variable \fIcdb_returnerror\fP to one
before calling \fIfb_opendb\fP, then the process will not terminate on errors.
In this case, \fIfb_opendb\fP will return either FB_AOK or FB_ERROR. The global
variable \fIcdb_error\fP will be set on FB_ERROR returns. See \fIsetup\fP(5)
for details on these and other global variables.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/libinit/opendb.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: opendb.c,v 9.0 2001/01/09 02:56:49 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Opendb_sid[] = "@(#) $Id: opendb.c,v 9.0 2001/01/09 02:56:49 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

static char *MSG1 = "creating new database";
extern short int cdb_opendb_level;
extern short int cdb_allow_links;
extern short int cdb_fixedwidth;
extern short int cdb_initscreen_done;
extern short int cdb_use_rpc;
extern short int cdb_returnerror;
extern short int cdb_error;
extern short int cdb_locklevel;
extern short int cdb_setup_done;   /* flag set after fb_setup() called */

extern int errno;

/*
 *  opendb - open a fb_database using given mode.
 *
 *	- if ixflag == NOINDEX, open index also.
 *	- if ixflag == WITHINDEX, open index also.
 *	- if ixflag == ALLINDEX, open index and autoindexes.
 *
 *      - if ixoption == OPTIONAL_INDEX, return
 *	- if ixoption == MAYBE_OPTIONAL_INDEX, fb_serror out
 *	- if ixoption == MUST_INDEX, fb_xerror out
 */

   fb_opendb(dp, mode, ixflag, ixoption)
      fb_database *dp;
      int mode, ixflag, ixoption;
      
      {
         char hname[FB_MAXNAME];
	 fb_database *save_dp;
	 int ignore_flag = 0, permissions = 0, simple_index = 0, st;

         if (cdb_t_lines == 0 || cdb_initscreen_done == 0)
	    cdb_batchmode = 1;
	 cdb_error = 0;
         errno = 0;
#if RPC
         if (cdb_use_rpc)
            return(fb_opendb_clnt(dp, mode, ixflag, ixoption));
#endif /* RPC */
         if (dp == NULL){
	    if (cdb_returnerror){
	       cdb_error = FB_BAD_DATA;
	       return(FB_ERROR);
	       }
	    fb_xerror(FB_BAD_DATA, SYSMSG[S_NO_INDEX], NIL);
	    }
         if (!cdb_setup_done)
	    fb_setup();		/* process /usr/lib/.cdbrc setup file */
         dp->dirty = CHAR_0;
	 if (cdb_opendb_level == 0){
	    if (cdb_locklevel <= 0)
	       ignore_flag = 1;
	    fb_initlock(ignore_flag, (fb_database *) dp);
	    }
         dp->fixedwidth = cdb_fixedwidth;
	 cdb_opendb_level++;
	 save_dp = cdb_db;
	 /* expand the dbase file name if needed */
         if ( (dp->dmap == NULL  || dp->dmap[0] == NULL)  ||
	      (dp->ddict == NULL || dp->ddict[0] == NULL)){
	    if (dp->dbase == NULL  || dp->dbase[0] == NULL){
	       if (cdb_returnerror){
		  cdb_error = FB_CANT_OPEN;
		  return(FB_ERROR);
		  }
	       fb_xerror(FB_CANT_OPEN, NIL, NIL);	/* panic if no name */
	       }
	    strcpy(hname, dp->dbase);
	    fb_dbargs(hname, NIL, dp);
	    }

	 /* expand the dbase index file name if needed */
         if (ixflag > 0)
            simple_index = 1;
	 if (ixflag > 0 && (dp->idict == NULL || dp->idict[0] == NULL)){
	    if ( (dp->dindex == NULL || dp->dindex[0] == NULL)){
	       /* error - no name to expand */
	       switch(ixoption){
	          case FB_MUST_INDEX:
		     if (cdb_returnerror){
			cdb_error = FB_CANT_OPEN;
			return(FB_ERROR);
			}
		     fb_xerror(FB_CANT_OPEN, NIL, NIL);
                     break;
		  case FB_MAYBE_OPTIONAL_INDEX:
		     if (cdb_returnerror){
			cdb_error = FB_CANT_OPEN;
			return(FB_ERROR);
			}
		     fb_serror(FB_CANT_OPEN, NIL, NIL);
		     break;
		  }
               simple_index = 0;
	       /* ixflag = 0; */
	       }
	    else{ /* expand index name */
	       strcpy(hname, dp->dindex);
	       fb_dbargs(NIL, hname, dp);
	       }
	    }
	    
         if (fb_getd_dict(dp) == FB_ERROR){
	    if (!cdb_batchmode)
	       fb_clear();
	    if (cdb_returnerror){
	       cdb_error = FB_BAD_DICT;
	       return(FB_ERROR);
	       }
	    fb_xerror(FB_BAD_DICT, dp->ddict, NIL);
	    }
         st = fb_getd_data(mode, dp);
         if (st == FB_LOCKED_ERROR){
            cdb_error = FB_LOCKED_ERROR;
            if (cdb_returnerror)
               return(FB_ERROR);
            fb_xerror(FB_CANT_OPEN, dp->dbase, "locked");
            }
         else if (st == FB_DIRTY_DBASE){
            cdb_error = FB_DIRTY_DBASE;
            return(FB_ERROR);
            }
         if (st == FB_ERROR){			/* attempt to create it */
	    if (!cdb_batchmode)
	       fb_clear();
	    if (access(dp->dbase, 0) != 0 && access(dp->dmap, 0) != 0 &&
	          (mode == READWRITE || mode == WRITE)){
	       if (!cdb_batchmode)
	          fb_fmessage(MSG1);
               fb_make_dbase(dp);
	       }
	    else {
               permissions = 0;
               if (mode == READ)
                  permissions = 4;
               else if (mode == READWRITE || mode == WRITE)
                  permissions = 6;
	       if (access(dp->dbase, permissions) != 0){
		  if (cdb_returnerror){
		     cdb_error = FB_CANT_OPEN;
		     return(FB_ERROR);
		     }
	          fb_xerror(FB_CANT_OPEN, dp->dbase, NIL);
		  }
	       if (access(dp->dmap, 6) != 0){	/* write needed for locks */
		  if (cdb_returnerror){
		     cdb_error = FB_CANT_OPEN;
		     return(FB_ERROR);
		     }
	          fb_xerror(FB_CANT_OPEN, dp->dmap, NIL);
		  }
	       /* this catches too many open files */
	       if (cdb_returnerror){
		  cdb_error = FB_CANT_OPEN;
		  return(FB_ERROR);
		  }
	       fb_xerror(FB_CANT_OPEN, dp->dbase, NIL);
	       }
	    }
	 if (ixflag > 1 && fb_getauto(dp, mode) == FB_ERROR){	/* ALLINDEX */
	    if (cdb_returnerror){
	       cdb_error = FB_BAD_INDEX;
	       return(FB_ERROR);
	       }
	    fb_xerror(FB_BAD_INDEX, dp->dbase, NIL);
	    }
	 if (simple_index > 0){				/* WITHINDEX */
	    strcpy(hname, dp->dindex);	/* save name for pos error msg */
	    if (fb_geti_dict(mode, dp) == FB_ERROR){
	       if (!cdb_batchmode)
		  fb_clear();
	       switch(ixoption){
	          case FB_MUST_INDEX:
		     if (cdb_returnerror){
			cdb_error = FB_BAD_DICT;
			return(FB_ERROR);
			}
		     fb_xerror(FB_BAD_DICT, dp->idict, NIL);
                     break;
		  case FB_MAYBE_OPTIONAL_INDEX:
		     break;
		  }
	       ixflag = 0;
	       }
	    else if (fb_geti_data(mode, dp) == FB_ERROR){
	       if (!cdb_batchmode)
		  fb_clear();
	       switch(ixoption){
	          case FB_MUST_INDEX:
		     if (cdb_returnerror){
			cdb_error = FB_CANT_OPEN;
			return(FB_ERROR);
			}
                     if (dp->b_tree && dp->b_seq != NULL){
                        fb_rootname(hname, dp->b_seq->bs_name);
		        fb_xerror(FB_BAD_INDEX, hname, NIL);
                        }
                     else
		        fb_xerror(FB_CANT_OPEN, hname, NIL);
                     break;
		  case FB_MAYBE_OPTIONAL_INDEX:
		     if (cdb_returnerror){
			cdb_error = FB_CANT_OPEN;
			return(FB_ERROR);
			}
		     fb_serror(FB_CANT_OPEN, hname, NIL);
                     break;
		  }
	       ixflag = 0;
	       }
	    }
	 if (ixflag == 0)
	    fb_emptyi_dict(dp);
         if (cdb_allow_links){
	    st = fb_initlink(dp);
            if (st == FB_ERROR){
               /*
                * if an FB_ERROR and NOT cdb_returnerror, fb_xerror will
                * have been already called in fb_initlink
                */
               if (cdb_returnerror){
                  cdb_error = FB_LINK_ERROR;
                  return(FB_ERROR);
                  }
	       fb_xerror(FB_LINK_ERROR, dp->ddict, NIL);
               }
            }
	 fb_nobuf(dp);			/* turn off buffering */
	 /* restore top dbase global variables  */
	 cdb_db = save_dp;
         if (cdb_db != NULL){
            cdb_kp = cdb_keymap = cdb_db->kp;
            cdb_ip = cdb_keyindx = cdb_db->ip;
            }
	 dp->refcnt++;
	 return(FB_AOK);
      }
